### PR TITLE
Milestone2 to Development

### PR DESCRIPTION
**Pull request by DickensJuma**
### What does this pr do?
-Build friendship feature for social media app
### Description of tasks to be completed?
1. Build friendship model, controller and view
2. Implement associations and validations
3. Write unit and integration tests

### CODE REVIEW

**Comment by sumancrest0001** [comment]
**Status: changes required :gear:
Hello Team,**
you have done a good job to implement some of the requested features for this milestone but some of them are still missing. Let's have a look at them and make changes accordingly.

[Required] Please make PR more descriptive.
[Required] Currently, friendship relationship is established for only one user but Whenever the request is accepted then two users should be friends of each other. That is not the case in your project. Please reference [this article](https://medium.com/@elizabethprendergast/using-custom-relation-queries-to-establish-friends-and-friendships-in-rails-and-activerecord-6c6e5825433a) [pathright-reference](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friend.png)
[Required] No tests are added. Please add tests to check the models, controllers, and other important methods. [pathright](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/tests.png)

[Required] we need to add 'add friend' button to request friendship. [pathright](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/send_invitation.png)

[Required] We need to see pending friend requests if any. The user can reject friend request also. [pathright](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/pending_request.png)

[Required] We need to add a request friendship button on both users list and user show pages. [pathright](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png)

[Required] Linter configuration is not working.

**Review by sumancrest0001 — CHANGES_REQUESTED** [review]
**Please make the necessary changes and submit the project for another coding review.
If you have any issues with the review, DM me on slack Suman Shrestha.
Happy coding !!!**

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi Team     
#### Great Job so far, everything works fine, but there are a few more things required. 
* [x] A `user` can invite friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/send_invitation.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [x] A `user` can see friendships requests ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/pending_request.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [x] A `user` has to accept friendship request to start friendship ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [x] A `user` can reject a friendship request ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [ ] There is a button to invite `user` to friendship on users list page and on single user page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* [x] Posts timeline should include posts written by logged-in user and their friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
- [ ] This `Pull Request` should be more descriptive,lets fix that. 
### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

File: app/views/users/index.html.erb [hunk 1]
```diff
@@ -2,11 +2,29 @@
...
+           <% if current_user.friend_invites(user) %>
+            <button class="btn btn-warning"> Pending Invitation </button>
+          <% elsif current_user.receive_invitation(user)%>
+            <%= link_to 'Accept', accept_request_path(user_id: user.id)%>
+            <%= link_to 'Decline', deny_request_path(user_id: user.id), method: :delete %>
+          <% elsif !current_user.friend?(user) && current_user.id != user.id %>
+            <%= link_to 'Add Friend', add_friend_path(user_id: user.id)%>
+            <% elsif current_user.friend?(user) %>
+            <%= link_to 'Remove friend', delete_friend_path(user_id: user.id), method: :delete, data: {confirm: "Are you sure you want to remove friend?"}, class: "btn btn-danger" %>
+          <% end %>
```
Comment (maelfosso) [inline app/views/users/index.html.erb]: That's called **logic in views**
Move that logic into views helpers

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,

Thank you for applying the suggested changes
However, there are still some missing requirements.


1. It's required, [read it here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png), that posts timeline includes posts written by logged-in user and their friends. It's not what you did.

2. **Remove Friend** feature is not working. Fix it please,

3. Wwrite integration tests for **friendships** feature please

4. Remove logic from views. Implement them into **views helpers**

Kindly make the suggested changes and submit a new code review request
Thanks

Mael FOSSO

**Review by stratospherique — APPROVED** [review]
### Great progress @team :clap: :clap: Congrats, the milestone is approved :heavy_check_mark: You may merge :+1: :+1: 

Keep up the good work :muscle: 
**_Ahmed Mahfoudh_**

**Pull request merged**